Add TruffleSOM support

owenoclee opened the pull request:
Basic support for TruffleSOM following on from [these proposed TruffleSOM changes](https://github.com/SOM-st/TruffleSOM/pull/24). Code completion, code navigation, and syntax error highlighting are (mostly) working but are not polished.

Commit:
Add first elements of a SOM language adapter

- introduce superclass for adapters
- adapt depending classes
- add SOM to the vscode extension
- add TruffleSOM as a dependency in .classpath

Commit:
Update SmalltalkAdapter to use latest truffle api

- require core-lib to be set
- use source coordinates for parse error messages

Commit:
Keep track of current SClass in TruffleSOM adapter

Commit:
Start working on code navigation/definition features

Commit:
Rename SmalltalkAdapter to TruffleSom

Commit:
Initial implementation of probes/gathering of additional parse information

- work on TruffleSomAdapter: class and instance methods now display in
  the outline and symbol search in VSCode

- use structural probes to get symbol info
- add navigation to instance/class fields
- add navigation to classes
- adjust navigation to methods
  Previously the navigation took you to inside the method. Now it takes you to the start of the method declaration.
- add workspace symbol search
  i.e. searches beginning with # in VSCode command palette

- use fields instead of symbols
  Due to the change in TruffleSOM to how field source sections are recorded

- implement code navigation
  - incl. for system classes
  - Bug: attempting to navigate to a primitive doesn't work yet due to no location info

- add code completion for TruffleSOM
  - not working for classes yet

Commit:
Rename Adapter to LanguageAdapter

smarr commented:
Thanks a lot for the PR.

Think this will take me a while longer.
There's a lot going on here, and I may want to push it just a little bit further.
Not yet sure exactly what to do.

Will look into the [TruffleSOM changes first](https://github.com/SOM-st/TruffleSOM/pull/24).

smarr commented:
First step for integration is done by adding SOM highlighting from the corresponding repo: #9.

smarr committed:
Use Invocation instead of Send interface

Signed-off-by: Stefan Marr <git@stefan-marr.de>

smarr committed:
Use SourceCoordinate from BD

Signed-off-by: Stefan Marr <git@stefan-marr.de>

smarr committed:
Use StructuralProbe from BD

Signed-off-by: Stefan Marr <git@stefan-marr.de>

smarr committed:
Symbols can be compared by identity

Signed-off-by: Stefan Marr <git@stefan-marr.de>

smarr committed:
Add TruffleSOM as submodule, update it, and adapt to changes

- fix constructor for TruffleSomCompiler
- added TODO
- adapt to change of how to access universe
- use .getFieldIndex directly, instead of via write node

Signed-off-by: Stefan Marr <git@stefan-marr.de>

smarr committed:
Added SOM tests

Signed-off-by: Stefan Marr <git@stefan-marr.de>

smarr committed:
Adapt build setup and update SOMns

Signed-off-by: Stefan Marr <git@stefan-marr.de>

smarr commented:
And CI is green, and even has two basic functionality tests :)

https://travis-ci.org/smarr/SOMns-vscode/builds/526145151

<img width="1423" alt="Screen Shot 2019-04-29 at 21 18 23" src="https://user-images.githubusercontent.com/40815/56924867-de2b6c00-6ac5-11e9-95a6-3b03699ce687.png">

smarr committed:
Rename SOMns/Newspeak specific classes and move to newspeak package

Signed-off-by: Stefan Marr <git@stefan-marr.de>

smarr committed:
Rename SOM related classes and move to .som package

Signed-off-by: Stefan Marr <git@stefan-marr.de>

smarr committed:
Move common code to LanguageAdapter

Signed-off-by: Stefan Marr <git@stefan-marr.de>

smarr committed:
Move fuzzy matching to new class

Reduces code duplication

Signed-off-by: Stefan Marr <git@stefan-marr.de>

smarr committed:
SOM does not have a vmMirror

Signed-off-by: Stefan Marr <git@stefan-marr.de>

smarr committed:
Remove dead code

Signed-off-by: Stefan Marr <git@stefan-marr.de>

smarr committed:
Keep language adapters as array

- allows to iterate over all adapters
- add explicit responsibility check based on uris to adapter
- avoid code duplication

Signed-off-by: Stefan Marr <git@stefan-marr.de>

smarr committed:
Disable assertion and document it

Signed-off-by: Stefan Marr <git@stefan-marr.de>

smarr committed:
Filter out self/blockSelf from displaying

Signed-off-by: Stefan Marr <git@stefan-marr.de>

smarr commented:
And another screenshot for documentation.

Both Truffle languages working in the same language server: SOM and SOMns

<img width="1247" alt="Screen Shot 2019-05-01 at 00 58 03" src="https://user-images.githubusercontent.com/40815/57000670-a5b28d80-6bac-11e9-8983-9affc9ca5e66.png">

smarr commented:
Finally merged. Thanks a lot for your work @owenoclee!
